// @composio/core/dist/fileUtils.node-Dq7ebZPC.mjs
import { a as ComposioSensitiveFilePathBlockedError, c as logger_default, i as ComposioFileUploadPathNotAllowedError, l as COMPOSIO_DIR, m as TEMP_FILES_DIRECTORY_NAME, t as ComposioFileNotFoundError } from "./FileModifierErrors-C7z16Dfi.mjs";
import { i as getRandomShortId, n as base64ToUint8Array, r as uint8ArrayToBase64 } from "./buffer-yOtEsT4J.mjs";
import { platform } from "#platform";
import fs from "node:fs";
import * as path from "node:path";
import crypto from "node:crypto";

//#region src/utils/sensitiveFileUploadPaths.node.ts
/**
* Blocks accidental upload of local files from well-known secret/credential locations
* during auto file upload (and {@link getFileDataAfterUploadingToS3}).
*/
/**
* Path segments (a single path component) that indicate a sensitive directory when
* they appear anywhere in a resolved local path. Compared case-insensitively on Windows.
*/
const BUILTIN_FILE_UPLOAD_PATH_DENY_SEGMENTS = [
	".ssh",
	".aws",
	".azure",
	".gnupg",
	".kube",
	".docker",
	".claude",
	".password-store",
	"keychains"
];
const SECRET_LIKE_BASENAME = /^(\.env(\.|$)|\.netrc$|\.pgpass$)/i;
/** Default SSH private key basenames (public keys like id_rsa.pub are allowed). */
const DEFAULT_PRIVATE_KEY_BASENAME = /^id_(rsa|ed25519|ecdsa|dsa|ecdsa_sk)(\.old)?$/i;
/**
* Returns normalized path segments, resolving symlinks when the path exists.
*/
function normalizePathSegments(filePath) {
	const absolute = path.resolve(filePath);
	let resolved = absolute;
	try {
		if (fs.existsSync(absolute)) resolved = fs.realpathSync(absolute);
	} catch {}
	return resolved.split(/[/\\]+/).filter(Boolean);
}
function getSensitiveFileUploadPathBlockReason(filePath, additionalDenySegments) {
	const segments = normalizePathSegments(filePath);
	const isWin = process.platform === "win32";
	const deny = new Set([...BUILTIN_FILE_UPLOAD_PATH_DENY_SEGMENTS, ...(additionalDenySegments ?? []).map((s) => s.trim()).filter(Boolean)].map((s) => isWin ? s.toLowerCase() : s));
	const segmentsForMatch = isWin ? segments.map((s) => s.toLowerCase()) : segments;
	for (let i = 0; i < segments.length; i++) if (deny.has(segmentsForMatch[i])) return `path segment "${segments[i]}" is in the sensitive file upload denylist`;
	const basename = segments.length > 0 ? segments[segments.length - 1] : "";
	if (basename) {
		if (SECRET_LIKE_BASENAME.test(basename) || DEFAULT_PRIVATE_KEY_BASENAME.test(basename)) return `file name "${basename}" looks like a credential, env, or private key file`;
		if (basename.toLowerCase() === "credentials") return "file name \"credentials\" is often used for cloud/API credential stores";
	}
	return null;
}
/**
* @throws {ComposioSensitiveFilePathBlockedError} if the path is not allowed
*/
function assertSafeFileUploadPath(filePath, options) {
	const reason = getSensitiveFileUploadPathBlockReason(filePath, options?.additionalDenySegments);
	if (reason) throw new ComposioSensitiveFilePathBlockedError(`Refusing to upload: ${reason}. To upload from this path anyway, set sensitiveFileUploadProtection: false on Composio (not recommended) or use a copy outside sensitive locations.`, { meta: {
		filePath,
		reason
	} });
}

//#endregion
//#region src/utils/uploadDirAllowlist.node.ts
/**
* Allowlist for automatic file upload during tool execution.
*
* This is only consulted when `dangerouslyAllowAutoUploadDownloadFiles` is enabled.
* Manual `composio.files.upload()` calls are not subject to the allowlist.
*
* - A local path is accepted iff its real (symlink-resolved, absolute) path is
*   inside one of the allowed directories on a path-component boundary.
*   `/tmp/foo` allows `/tmp/foo/bar` but NOT `/tmp/foo-bar`.
* - URLs (`http(s)://…`) and `File`/`Blob` values never hit this check.
* - User-provided `fileUploadDirs` fully replaces the default `[~/.composio/temp]`.
* - On Windows, path comparisons are case-insensitive.
*/
function normalizeForCompare(p) {
	return process.platform === "win32" ? p.toLowerCase() : p;
}
/**
* True if `child` is equal to `parent` or nested inside it on a component
* boundary. Assumes both paths are absolute and normalized.
*/
function isInsideDir(child, parent) {
	const a = normalizeForCompare(child);
	const b = normalizeForCompare(parent);
	if (a === b) return true;
	const sep = path.sep;
	const bWithSep = b.endsWith(sep) ? b : b + sep;
	return a.startsWith(bWithSep);
}
/**
* Returns the real (symlink-resolved) absolute path when the target exists,
* else `null`. Mirrors the semantics of the sensitive-path util.
*/
function tryRealpath(absPath) {
	try {
		if (fs.existsSync(absPath)) return fs.realpathSync(absPath);
	} catch {}
	return null;
}
function formatAllowlist(dirs) {
	if (dirs.length === 0) return "  (none configured — all local paths are blocked)";
	return dirs.map((d) => `  - ${d}`).join("\n");
}
function buildHelpFooter(allowlist) {
	return [
		"",
		"Allowed upload directories (fileUploadDirs):",
		formatAllowlist(allowlist),
		"",
		"Fix one of the following:",
		"",
		"  1. Recommended — move the file under an allowed directory, or add your",
		"     directory to the allowlist:",
		"       new Composio({",
		"         fileUploadDirs: ['/abs/path/to/uploads', '~/.composio/temp'],",
		"       })",
		"     Note: user-provided `fileUploadDirs` REPLACES the default",
		"     `~/.composio/temp`. Include it explicitly if you want staged uploads",
		"     to keep working.",
		"",
		"  2. Pass the file by URL or as a File/Blob object instead of a filesystem",
		"     path. URLs and File objects are never path-checked.",
		"",
		"  3. (Dangerous) Bypass the allowlist entirely by opting into automatic",
		"     file upload/download from any readable path:",
		"       new Composio({ dangerouslyAllowAutoUploadDownloadFiles: true })",
		"     WARNING: This lets a prompt-injected or misbehaving tool read ANY",
		"     file the process can access. Only enable it in trusted, sandboxed",
		"     environments. The built-in sensitive-path denylist (.ssh, .aws,",
		"     .env, private SSH keys, etc.) still applies unless you also set",
		"     `sensitiveFileUploadProtection: false`."
	].join("\n");
}
/**
* @throws {ComposioFileNotFoundError} when `filePath` does not exist on disk.
* @throws {ComposioFileUploadPathNotAllowedError} when the resolved path is
*         outside every entry of `allowlist`.
*/
function assertPathInsideUploadDirs(filePath, allowlist) {
	const attempted = filePath;
	const abs = path.resolve(filePath);
	const real = tryRealpath(abs);
	if (!real) {
		const cwd = process.cwd();
		const parent = path.dirname(abs);
		const parentExists = (() => {
			try {
				return fs.existsSync(parent);
			} catch {
				return false;
			}
		})();
		throw new ComposioFileNotFoundError([
			`Refusing to auto-upload "${attempted}": the file does not exist on disk.`,
			"",
			`Path attempted:   ${attempted}`,
			`Resolved to:      ${abs}`,
			`Process cwd:      ${cwd}`,
			`Parent exists:    ${parentExists ? "yes (" + parent + ")" : "no (" + parent + ")"}`,
			"",
			"Common causes:",
			"  - Typo in the filename passed to the tool.",
			"  - Relative path resolved against the wrong working directory",
			"    (relative paths use process.cwd() at the moment of upload).",
			"  - File was deleted between the tool being called and the upload starting.",
			"",
			buildHelpFooter(allowlist)
		].join("\n"), { meta: {
			attempted,
			resolved: abs,
			cwd,
			allowlist
		} });
	}
	if (allowlist.length === 0) throw new ComposioFileUploadPathNotAllowedError([
		`Refusing to auto-upload "${attempted}": no upload directories are configured.`,
		"",
		`Path attempted:   ${attempted}`,
		`Resolved to:      ${real}`,
		"",
		"Automatic file upload during tool execution is locked down by default",
		"to prevent a prompt-injected tool from exfiltrating server files",
		"(source code, .env, SSH keys, etc.).",
		buildHelpFooter(allowlist)
	].join("\n"), { meta: {
		attempted,
		resolved: real,
		allowlist
	} });
	for (const dir of allowlist) if (isInsideDir(real, tryRealpath(dir) ?? path.resolve(dir))) return;
	throw new ComposioFileUploadPathNotAllowedError([
		`Refusing to auto-upload "${attempted}": resolved path is not inside any`,
		"directory in the configured `fileUploadDirs` allowlist.",
		"",
		`Path attempted:   ${attempted}`,
		`Resolved to:      ${real}`,
		buildHelpFooter(allowlist)
	].join("\n"), { meta: {
		attempted,
		resolved: real,
		allowlist
	} });
}

//#endregion
//#region src/utils/fileUtils.node.ts
const getExtensionFromMimeType = (mimeType) => {
	const mimeToExt = {
		"text/plain": "txt",
		"text/html": "html",
		"text/css": "css",
		"text/javascript": "js",
		"application/json": "json",
		"application/xml": "xml",
		"application/pdf": "pdf",
		"application/zip": "zip",
		"application/x-zip-compressed": "zip",
		"application/gzip": "gz",
		"application/x-tar": "tar",
		"application/msword": "doc",
		"application/vnd.openxmlformats-officedocument.wordprocessingml.document": "docx",
		"application/vnd.ms-excel": "xls",
		"application/vnd.openxmlformats-officedocument.spreadsheetml.sheet": "xlsx",
		"application/vnd.ms-powerpoint": "ppt",
		"application/vnd.openxmlformats-officedocument.presentationml.presentation": "pptx",
		"image/jpeg": "jpg",
		"image/jpg": "jpg",
		"image/png": "png",
		"image/gif": "gif",
		"image/svg+xml": "svg",
		"image/webp": "webp",
		"image/bmp": "bmp",
		"image/tiff": "tiff",
		"audio/mpeg": "mp3",
		"audio/wav": "wav",
		"audio/ogg": "ogg",
		"video/mp4": "mp4",
		"video/mpeg": "mpeg",
		"video/quicktime": "mov",
		"video/x-msvideo": "avi",
		"video/webm": "webm"
	};
	const cleanMimeType = mimeType.split(";")[0].toLowerCase().trim();
	if (mimeToExt[cleanMimeType]) return mimeToExt[cleanMimeType];
	const parts = cleanMimeType.split("/");
	if (parts.length === 2) {
		const cleanSubtype = parts[1].toLowerCase();
		if (cleanSubtype.includes("+")) {
			const plusParts = cleanSubtype.split("+");
			const prefix = plusParts[0];
			const suffix = plusParts[plusParts.length - 1];
			if ([
				"svg",
				"atom",
				"rss"
			].includes(prefix)) return prefix;
			if ([
				"json",
				"xml",
				"yaml",
				"zip",
				"gzip"
			].includes(suffix)) return suffix;
			return suffix;
		}
		return cleanSubtype || "txt";
	}
	return "txt";
};
const generateTimestampedFilename = (extension, prefix) => {
	return `${prefix || "file_ts"}${Date.now()}${getRandomShortId()}.${extension}`;
};
const readFileContent = async (filePath) => {
	try {
		if (!platform.supportsFileSystem) throw new Error("File system operations are not supported in this runtime environment");
		const content = platform.readFileSync(filePath);
		return {
			fileName: generateTimestampedFilename(filePath.split(".").pop() || "txt"),
			content: content instanceof Uint8Array ? uint8ArrayToBase64(content) : uint8ArrayToBase64(new TextEncoder().encode(content)),
			mimeType: "application/octet-stream"
		};
	} catch (error) {
		throw new Error(`Error reading file at ${filePath}: ${error}`);
	}
};
const readFileContentFromURL = async (path$1) => {
	const response = await fetch(path$1);
	if (!response.ok) throw new Error(`Failed to fetch file: ${response.statusText}`);
	const arrayBuffer = await response.arrayBuffer();
	const content = new Uint8Array(arrayBuffer);
	const mimeType = response.headers.get("content-type") || "application/octet-stream";
	const pathname = new URL(path$1).pathname;
	let fileName = platform.basename(pathname);
	if (!fileName || fileName === "/") fileName = generateTimestampedFilename(getExtensionFromMimeType(mimeType));
	else if (!fileName.includes(".")) fileName = generateTimestampedFilename(getExtensionFromMimeType(mimeType));
	return {
		content: uint8ArrayToBase64(content),
		mimeType,
		fileName
	};
};
const uploadFileToS3 = async (fileName, content, toolSlug, toolkitSlug, mimeType, client) => {
	const contentBytes = base64ToUint8Array(content);
	const { key, new_presigned_url: signedURL } = await client.files.createPresignedURL({
		filename: fileName,
		mimetype: mimeType,
		md5: crypto.createHash("md5").update(contentBytes).digest("hex"),
		tool_slug: toolSlug,
		toolkit_slug: toolkitSlug
	});
	logger_default.debug(`Uploading ${key} file to S3: ${key}`);
	const uploadBuffer = new Uint8Array(contentBytes.byteLength);
	uploadBuffer.set(contentBytes);
	const uploadResponse = await fetch(signedURL, {
		method: "PUT",
		body: uploadBuffer,
		headers: {
			"Content-Type": mimeType,
			"Content-Length": contentBytes.length.toString()
		}
	});
	if (!uploadResponse.ok) throw new Error(`Failed to upload file to S3: ${uploadResponse.statusText}`);
	return key;
};
const readFile = async (file) => {
	if (file instanceof File) {
		const content = await file.arrayBuffer();
		return {
			fileName: file.name,
			content: uint8ArrayToBase64(new Uint8Array(content)),
			mimeType: file.type
		};
	} else if (typeof file === "string") if (file.startsWith("http")) return await readFileContentFromURL(file);
	else return await readFileContent(file);
	throw new Error("Invalid file type");
};
const getFileDataAfterUploadingToS3 = async (file, { toolSlug, toolkitSlug, client, sensitiveFileUploadProtection, fileUploadPathDenySegments, fileUploadAllowlist }) => {
	if (!file) throw new Error("Either path or blob must be provided");
	const isLocalPath = typeof file === "string" && !file.startsWith("http");
	if (isLocalPath && fileUploadAllowlist !== void 0) assertPathInsideUploadDirs(file, fileUploadAllowlist);
	if (sensitiveFileUploadProtection !== false && isLocalPath) assertSafeFileUploadPath(file, { additionalDenySegments: fileUploadPathDenySegments });
	const fileData = await readFile(file);
	logger_default.debug(`Uploading file to S3...`);
	const s3key = await uploadFileToS3(platform.basename(fileData.fileName), fileData.content, toolSlug, toolkitSlug, fileData.mimeType, client);
	logger_default.debug(`Done! File uploaded to S3: ${s3key}`, JSON.stringify(fileData, null, 2));
	return {
		name: fileData.fileName,
		mimetype: fileData.mimeType,
		s3key
	};
};
const downloadFileFromS3 = async ({ toolSlug, s3Url, mimeType, fileDownloadDir }) => {
	const response = await fetch(s3Url);
	if (!response.ok) throw new Error(`Failed to download file: ${response.statusText}`);
	const data = await response.arrayBuffer();
	const fileName = generateTimestampedFilename(getExtensionFromMimeType(mimeType), `${toolSlug}_`);
	return {
		name: fileName,
		mimeType,
		s3Url,
		filePath: saveFile(fileName, new Uint8Array(data), {
			isTempFile: fileDownloadDir === void 0,
			outputDir: fileDownloadDir
		})
	};
};
/**
* Gets the Composio directory.
* @param createDirIfNotExists - Whether to create the directory if it doesn't exist.
* @returns The path to the Composio directory.
*/
const getComposioDir = (createDirIfNotExists = false) => {
	try {
		const homeDir = platform.homedir();
		if (!homeDir) return null;
		const composioDir = platform.joinPath(homeDir, COMPOSIO_DIR);
		if (createDirIfNotExists && platform.supportsFileSystem && !platform.existsSync(composioDir)) platform.mkdirSync(composioDir);
		return composioDir;
	} catch (_error) {
		return null;
	}
};
/**
* Gets the Composio temporary files directory.
* @param createDirIfNotExists - Whether to create the directory if it doesn't exist.
* @returns The path to the Composio temporary files directory.
*/
const getComposioTempFilesDir = (createDirIfNotExists = false) => {
	try {
		const homeDir = platform.homedir();
		if (!homeDir) return null;
		const composioFilesDir = platform.joinPath(homeDir, COMPOSIO_DIR, TEMP_FILES_DIRECTORY_NAME);
		if (createDirIfNotExists && platform.supportsFileSystem && !platform.existsSync(composioFilesDir)) platform.mkdirSync(composioFilesDir);
		return composioFilesDir;
	} catch (_error) {
		return null;
	}
};
/**
* Saves a file to a local directory.
*
* Directory precedence (highest first):
*   1. `options.outputDir` (explicit, absolute or relative; created if missing)
*   2. `options.isTempFile === true` → `<home>/.composio/files`
*   3. fallback → `<home>/.composio`
*
* @returns The absolute path to the saved file, or `null` if the runtime has
*          no filesystem (e.g. Cloudflare Workers) or the save failed.
*/
const saveFile = (file, content, options = {}) => {
	const opts = typeof options === "boolean" ? { isTempFile: options } : options;
	try {
		if (!platform.supportsFileSystem) {
			logger_default.debug("File system operations are not supported in this runtime environment");
			return null;
		}
		let composioFilesDir;
		if (opts.outputDir) {
			const absDir = platform.joinPath(opts.outputDir);
			try {
				if (!platform.existsSync(absDir)) platform.mkdirSync(absDir);
				composioFilesDir = absDir;
			} catch (err) {
				logger_default.warn(`fileDownloadDir "${absDir}" is not writable, falling back to the default (~/.composio/files). Error: ${err}`);
				composioFilesDir = getComposioTempFilesDir(true);
			}
		} else composioFilesDir = opts.isTempFile ? getComposioTempFilesDir(true) : getComposioDir(true);
		if (!composioFilesDir) return null;
		const filePath = platform.joinPath(composioFilesDir, platform.basename(file));
		logger_default.info(`Saving file to: ${filePath}`);
		if (content instanceof Uint8Array) platform.writeFileSync(filePath, content);
		else platform.writeFileSync(filePath, content, "utf8");
		return filePath;
	} catch (_error) {
		logger_default.debug(`Error saving file: ${_error}`);
		return null;
	}
};

//#endregion
export { getFileDataAfterUploadingToS3 as n, downloadFileFromS3 as t };
