// @composio/core/dist/FileModifierErrors-C7z16Dfi.mjs
import { t as __exportAll } from "./chunk-CWQeghP5.mjs";
import { BadRequestError } from "@composio/client";
import { ZodError } from "zod/v3";
import chalk from "chalk";

//#region src/utils/env.ts
const getEnvVariable = (name, defaultValue = void 0) => {
	try {
		return process.env[name] || defaultValue;
	} catch (_e) {
		return defaultValue;
	}
};
/**
* Method to return all the envs based on a specific prefix
* @param prefix
* @example
* COMPOSIO_TOOLKIT_VERSION_GITHUB=20250902_00
* COMPOSIO_TOOLKIT_VERSION_SLACK=20250902_00
* COMPOSIO_TOOLKIT_VERSION_GMAIL=latest
* @returns {Record<string, unknown>}
* @returns
*/
const getEnvsWithPrefix = (prefix) => {
	try {
		if (process && process.env) return Object.fromEntries(Object.entries(process.env).filter(([key]) => key.startsWith(prefix)));
		else return {};
	} catch (error) {
		return {};
	}
};

//#endregion
//#region src/utils/constants.ts
var constants_exports = /* @__PURE__ */ __exportAll({
	CLIENT_PUSHER_KEY: () => CLIENT_PUSHER_KEY,
	COMPOSIO_DIR: () => COMPOSIO_DIR,
	COMPOSIO_LOG_LEVEL: () => COMPOSIO_LOG_LEVEL,
	DEFAULT_BASE_URL: () => DEFAULT_BASE_URL,
	DEFAULT_TOOL_ROUTER_SESSION_FILES_MOUNT_ID: () => DEFAULT_TOOL_ROUTER_SESSION_FILES_MOUNT_ID,
	DEFAULT_WEB_URL: () => DEFAULT_WEB_URL,
	IS_DEVELOPMENT_OR_CI: () => IS_DEVELOPMENT_OR_CI,
	TELEMETRY_URL: () => TELEMETRY_URL,
	TEMP_FILES_DIRECTORY_NAME: () => TEMP_FILES_DIRECTORY_NAME,
	USER_DATA_FILE_NAME: () => USER_DATA_FILE_NAME
});
const COMPOSIO_DIR = ".composio";
const USER_DATA_FILE_NAME = "user_data.json";
const TEMP_FILES_DIRECTORY_NAME = "files";
const DEFAULT_BASE_URL = "https://backend.composio.dev";
const DEFAULT_WEB_URL = "https://dashboard.composio.dev/";
const TELEMETRY_URL = "https://app.composio.dev";
const CLIENT_PUSHER_KEY = getEnvVariable("CLIENT_PUSHER_KEY") || "ff9f18c208855d77a152";
const COMPOSIO_LOG_LEVEL = getEnvVariable("COMPOSIO_LOG_LEVEL");
const IS_DEVELOPMENT_OR_CI = getEnvVariable("DEVELOPMENT") || getEnvVariable("CI") || false;
const DEFAULT_TOOL_ROUTER_SESSION_FILES_MOUNT_ID = "files";

//#endregion
//#region src/utils/logger.ts
const LOG_LEVELS = {
	silent: -1,
	error: 0,
	warn: 1,
	info: 2,
	debug: 3
};
/**
* Get the current log level from environment variables.
* Defaults to 'info' if not set or invalid.
* @returns {LogLevel} The current log level
*/
const getLogLevel = () => {
	const envLevel = (COMPOSIO_LOG_LEVEL ?? "info")?.toLowerCase();
	return envLevel && envLevel in LOG_LEVELS ? envLevel : "info";
};
var Logger = class {
	level;
	includeTimestamp;
	console;
	constructor(options = {}) {
		this.level = options.level ?? getLogLevel();
		this.includeTimestamp = options.includeTimestamp ?? true;
		this.console = console;
	}
	formatMessage(args) {
		const formattedArgs = args.map((arg, index) => {
			if (typeof arg === "object") return JSON.stringify(arg);
			else {
				if (index === 0) if (args.length > 1) return chalk.yellow(`${arg}`);
				else return String(arg);
				return String(arg);
			}
		}).join("\n");
		if (!this.includeTimestamp) return formattedArgs;
		const timestamp = (/* @__PURE__ */ new Date()).toISOString();
		return `${chalk.gray(timestamp)} - ${formattedArgs}`;
	}
	shouldLog(level) {
		return LOG_LEVELS[level] <= LOG_LEVELS[this.level];
	}
	error(...args) {
		if (this.shouldLog("error")) this.console.error(this.formatMessage(args));
	}
	warn(...args) {
		if (this.shouldLog("warn")) this.console.warn(this.formatMessage(args));
	}
	info(...args) {
		if (this.shouldLog("info")) this.console.info(this.formatMessage(args));
	}
	debug(...args) {
		if (this.shouldLog("debug")) this.console.debug(this.formatMessage(args));
	}
};
const logger = new Logger();
var logger_default = logger;

//#endregion
//#region src/errors/ComposioError.ts
/**
* @fileoverview Error class for Composio.
*
* This error class is used to create custom errors for Composio.
* It extends the built-in Error class and adds additional properties for code, status code, cause, meta, and possible fixes.
*
* @author Musthaq Ahamad <musthaq@composio.dev>
* @module errors/ComposioError
* @see {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error}
*/
/**
* Base error class for all Composio errors
*/
var ComposioError$1 = class ComposioError$1 extends Error {
	/** @readonly Error name */
	name = "ComposioError";
	code;
	possibleFixes;
	errorId;
	/**
	* Creates a new ComposioError
	* @param message Error message
	* @param options Additional error options
	*/
	constructor(message, options = {}) {
		super(message);
		if (Error.captureStackTrace) Error.captureStackTrace(this, this.constructor);
		const statusCode = options.statusCode || (options.cause instanceof BadRequestError ? options.cause.status : void 0);
		this.code = `TS-SDK::${options.code}`;
		this.possibleFixes = options.possibleFixes;
		this.definePropertyIfExists("statusCode", statusCode);
		this.definePropertyIfExists("cause", options.cause);
		const combinedStack = options.cause instanceof Error ? ComposioError$1.combineStackTraces(options.cause.stack, this.stack) : options.stack ?? this.stack;
		this.definePropertyIfExists("stack", combinedStack);
		if (options.meta && Object.keys(options.meta).length > 0) this.definePropertyIfExists("meta", options.meta);
	}
	/**
	* Helper method to define a property only if it has a value
	* @param propertyName Name of the property to define
	* @param value Value to assign to the property
	* @private
	*/
	definePropertyIfExists(propertyName, value) {
		if (value !== void 0) Object.defineProperty(this, propertyName, {
			value,
			enumerable: true,
			writable: false,
			configurable: true
		});
	}
	/**
	* Helper method to combine stack traces when wrapping errors
	* This ensures the full call chain is preserved
	* @param originalStack The stack of the error being wrapped
	* @param currentStack The stack of the wrapper error
	* @returns Combined stack trace
	* @private
	*/
	static combineStackTraces(originalStack, currentStack) {
		if (!originalStack) return currentStack;
		if (!currentStack) return originalStack;
		const currentHeader = currentStack.split("\n")[0];
		const originalStackBody = originalStack.split("\n").slice(1).join("\n");
		return `${currentHeader}\n${currentStack.split("\n").slice(1).join("\n")}\n\nCaused by:\n${originalStackBody}`;
	}
	/**
	* Extract and normalize error data for formatting
	* @param includeStack Whether to include stack trace information
	* @returns Structured error data for formatting
	* @private
	*/
	getErrorData(includeStack = false) {
		const data = {
			name: this.name,
			message: this.message
		};
		const { cause, code, stack, statusCode, meta, possibleFixes } = this;
		if (cause !== void 0) {
			const rawCause = cause;
			data.cause = rawCause instanceof Error ? rawCause.message : String(rawCause);
		}
		if (code) data.code = code;
		if (statusCode !== void 0) data.statusCode = statusCode;
		if (meta) data.meta = meta;
		if (possibleFixes) data.possibleFixes = possibleFixes;
		if (includeStack && stack) if (stack.includes("Caused by:")) {
			const [currentStack, causeStack] = stack.split("Caused by:");
			data.stack = [
				...currentStack.split("\n").slice(1),
				"Caused by:",
				...causeStack.split("\n")
			];
		} else data.stack = stack.split("\n").slice(1);
		return data;
	}
	/**
	* Prints a user-friendly, colorful representation of the error to the logger
	* @param includeStack Whether to include the stack trace in the output (default: false)
	*/
	prettyPrint(includeStack = false) {
		const data = this.getErrorData(includeStack);
		let output = "\n" + chalk.bgRed.white.bold(" ERROR ") + " " + chalk.white.bold(data.message) + "\n";
		if (data.code) output += chalk.yellow(`Error Code: ${data.code}`) + "\n";
		if (data.statusCode !== void 0) output += chalk.yellow(`Status: ${data.statusCode}`) + "\n";
		if (data.cause) {
			output += chalk.gray("Reason:") + "\n";
			output += "  " + chalk.white(data.cause) + "\n";
		}
		if (data.meta) {
			output += chalk.gray("Additional Information:") + "\n";
			output += "  " + chalk.white(JSON.stringify(data.meta, null, 2).replace(/\n/g, "\n  ")) + "\n";
		}
		if (data.possibleFixes?.length) {
			output += "\n" + chalk.cyan.bold("Try the following:") + "\n";
			const fixes = data.possibleFixes?.map((fix, index) => ` ${index + 1}. ` + chalk.white(fix));
			output += fixes?.join("\n") + "\n";
		}
		if (data.stack?.length) {
			output += "\n" + chalk.gray("Stack Trace:") + "\n";
			output += chalk.gray(data.stack.join("\n")) + "\n";
		}
		output += "\n";
		logger_default.error(output);
	}
	/**
	* Static factory method to create and pretty print an error in one step
	* @param message Error message
	* @param options Error options
	* @param includeStack Whether to include the stack trace in the output
	* @returns The created error instance
	*/
	static createAndPrint(message, options = {}, includeStack = false) {
		const error = new ComposioError$1(message, options);
		error.prettyPrint(includeStack);
		return error;
	}
	/**
	* Utility function to handle errors in a consistent way
	* This properly displays the error without throwing
	* @param error The error to handle
	* @param options Options for error handling
	*/
	static handle(error, options = {}) {
		const { includeStack = false, exitProcess = false } = options;
		if (error instanceof ComposioError$1) error.prettyPrint(includeStack);
		else if (error instanceof ZodError) this.handleZodError(error, includeStack);
		else if (error instanceof Error) this.handleStandardError(error, includeStack);
		else this.handleUnknownError(error);
		if (exitProcess) this.throwError(error);
	}
	/**
	* Utility function to handle errors and then throw them
	* This properly displays the error and then throws it, allowing callers to catch it.
	* Use this for fatal errors that should stop execution.
	* @param error The error to handle and throw
	* @param includeStack Whether to include the stack trace in the output
	* @throws ComposioError - Always throws after displaying the error
	*/
	static handleAndThrow(error, includeStack = false) {
		this.handle(error, { includeStack });
		this.throwError(error);
	}
	/**
	* Helper method to throw an error as a ComposioError
	* @param error The error to throw
	* @private
	*/
	static throwError(error) {
		if (error instanceof ComposioError$1) throw error;
		else if (error instanceof Error) throw new ComposioError$1(error.message, { cause: error });
		else throw new ComposioError$1(String(error));
	}
	/**
	* Helper method to handle Zod validation errors
	* @param error The Zod error to handle
	* @param includeStack Whether to include the stack trace
	* @private
	*/
	static handleZodError(error, includeStack) {
		logger_default.error("\n" + chalk.bgRed.white.bold(" ERROR ") + " " + chalk.white.bold(error.message));
		logger_default.error(chalk.gray("Invalid parameters:"));
		error.errors.forEach((err) => {
			logger_default.error(chalk.yellow(err.path.join(".")) + " " + chalk.white(err.message));
		});
		logger_default.error(chalk.gray("Expected parameters:"));
		error.errors.forEach((err) => {
			logger_default.error(chalk.yellow(err.path.join(".")) + " " + chalk.white(err.message));
		});
		if (includeStack) {
			logger_default.error("\n" + chalk.gray("Validation Errors:"));
			error.errors.forEach((err) => {
				const path = err.path.join(".");
				logger_default.error(chalk.gray("  • ") + chalk.yellow(path ? `${path}: ` : "") + chalk.white(err.message));
			});
			if (error.stack) {
				logger_default.error("\n" + chalk.gray("Stack Trace:"));
				const stackLines = error.stack.split("\n").slice(1);
				logger_default.error(chalk.gray(stackLines.join("\n")));
			}
		}
		logger_default.error("");
	}
	/**
	* Helper method to handle standard Error objects
	* @param error The standard error to handle
	* @param includeStack Whether to include the stack trace
	* @private
	*/
	static handleStandardError(error, includeStack) {
		logger_default.error("\n" + chalk.bgRed.white.bold(" ERROR ") + " " + chalk.white.bold(error.message));
		if (includeStack && error.stack) {
			logger_default.error("\n" + chalk.gray("Stack Trace:"));
			const stackLines = error.stack.split("\n").slice(1);
			logger_default.error(chalk.gray(stackLines.join("\n")));
		}
		logger_default.error("");
	}
	/**
	* Helper method to handle unknown error types
	* @param error The unknown error value
	* @private
	*/
	static handleUnknownError(error) {
		logger_default.error("\n" + chalk.bgRed.white.bold(" ERROR ") + " " + chalk.white.bold("Unknown error occurred"));
		if (error !== null && error !== void 0) {
			logger_default.error(chalk.gray("Error details:"));
			logger_default.error("  " + chalk.white(String(error)));
		}
		logger_default.error("");
	}
};

//#endregion
//#region src/errors/FileModifierErrors.ts
const FileModifierErrorCodes = {
	FILE_UPLOAD_FAILED: "FILE_UPLOAD_FAILED",
	SENSITIVE_FILE_PATH_BLOCKED: "SENSITIVE_FILE_PATH_BLOCKED",
	FILE_UPLOAD_ABORTED: "FILE_UPLOAD_ABORTED",
	FILE_UPLOAD_PATH_NOT_ALLOWED: "FILE_UPLOAD_PATH_NOT_ALLOWED",
	FILE_NOT_FOUND: "FILE_NOT_FOUND"
};
var ComposioFileUploadError = class extends ComposioError$1 {
	constructor(message = "Failed to upload file", options = {}) {
		super(message, {
			...options,
			code: options.code ?? FileModifierErrorCodes.FILE_UPLOAD_FAILED,
			possibleFixes: options.possibleFixes || ["Check if the file exists in the location provided"]
		});
		this.name = "ComposioFileUploadError";
	}
};
/**
* Thrown when a local file path is refused before upload (sensitive directory or
* credential-like file name). Subclass of {@link ComposioFileUploadError} so existing
* `instanceof ComposioFileUploadError` handlers still apply.
*/
var ComposioSensitiveFilePathBlockedError = class extends ComposioFileUploadError {
	constructor(message, options = {}) {
		super(message, {
			...options,
			code: FileModifierErrorCodes.SENSITIVE_FILE_PATH_BLOCKED,
			possibleFixes: options.possibleFixes ?? [
				"Use a file outside ~/.ssh, ~/.aws, and other credential locations",
				"Set sensitiveFileUploadProtection: false only if you understand the risk",
				"Add path segments to block with fileUploadPathDenySegments"
			]
		});
		this.name = "ComposioSensitiveFilePathBlockedError";
	}
};
/**
* Thrown when a `beforeFileUpload` hook returns `false`.
*/
var ComposioFileUploadAbortedError = class extends ComposioFileUploadError {
	constructor(message, options = {}) {
		super(message, {
			...options,
			code: FileModifierErrorCodes.FILE_UPLOAD_ABORTED,
			possibleFixes: options.possibleFixes ?? ["Change beforeFileUpload to return a file path to upload, or allow the upload to proceed"]
		});
		this.name = "ComposioFileUploadAbortedError";
	}
};
/**
* Thrown when automatic file upload during tool execution is attempted from a path
* that is outside the configured {@link ComposioConfig.fileUploadDirs} allowlist.
*
* This only fires for auto-upload (enabled via
* `dangerouslyAllowAutoUploadDownloadFiles: true`). Manual `composio.files.upload()`
* calls are never subject to this check.
*/
var ComposioFileUploadPathNotAllowedError = class extends ComposioFileUploadError {
	constructor(message, options = {}) {
		super(message, {
			...options,
			code: FileModifierErrorCodes.FILE_UPLOAD_PATH_NOT_ALLOWED,
			possibleFixes: options.possibleFixes ?? [
				"Move the file under one of the directories in `fileUploadDirs`",
				"Stage the file under the default upload dir (`~/.composio/temp`)",
				"Add the containing directory to `fileUploadDirs` when constructing Composio",
				"Pass the file by URL or as a File/Blob object instead of a filesystem path"
			]
		});
		this.name = "ComposioFileUploadPathNotAllowedError";
	}
};
/**
* Thrown when a local file path passed for upload does not exist on disk.
* Provides extra debugging context (resolved absolute path, cwd, allowlist state)
* to help the caller distinguish between "typo" and "not-allowlisted" failures.
*/
var ComposioFileNotFoundError = class extends ComposioFileUploadError {
	constructor(message, options = {}) {
		super(message, {
			...options,
			code: FileModifierErrorCodes.FILE_NOT_FOUND,
			possibleFixes: options.possibleFixes ?? [
				"Double-check the path for typos",
				"Remember relative paths are resolved against process.cwd() at upload time",
				"Verify the file exists and the process has read permission"
			]
		});
		this.name = "ComposioFileNotFoundError";
	}
};

//#endregion
export { getEnvVariable as _, ComposioSensitiveFilePathBlockedError as a, logger_default as c, DEFAULT_BASE_URL as d, DEFAULT_TOOL_ROUTER_SESSION_FILES_MOUNT_ID as f, constants_exports as g, USER_DATA_FILE_NAME as h, ComposioFileUploadPathNotAllowedError as i, COMPOSIO_DIR as l, TEMP_FILES_DIRECTORY_NAME as m, ComposioFileUploadAbortedError as n, FileModifierErrorCodes as o, IS_DEVELOPMENT_OR_CI as p, ComposioFileUploadError as r, ComposioError$1 as s, ComposioFileNotFoundError as t, COMPOSIO_LOG_LEVEL as u, getEnvsWithPrefix as v };
