// @composio/client/src/resources/logs/logs.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../../core/resource';
import * as ToolsAPI from './tools';
import { ToolListParams, ToolListResponse, ToolRetrieveResponse, Tools } from './tools';
import * as TriggersAPI from './triggers';
import { TriggerListParams, TriggerListResponse, TriggerRetrieveResponse, Triggers } from './triggers';

export class Logs extends APIResource {
  triggers: TriggersAPI.Triggers = new TriggersAPI.Triggers(this._client);
  tools: ToolsAPI.Tools = new ToolsAPI.Tools(this._client);
}

Logs.Triggers = Triggers;
Logs.Tools = Tools;

export declare namespace Logs {
  export {
    Triggers as Triggers,
    type TriggerRetrieveResponse as TriggerRetrieveResponse,
    type TriggerListResponse as TriggerListResponse,
    type TriggerListParams as TriggerListParams,
  };

  export {
    Tools as Tools,
    type ToolRetrieveResponse as ToolRetrieveResponse,
    type ToolListResponse as ToolListResponse,
    type ToolListParams as ToolListParams,
  };
}
