// @composio/client/resources/connected-accounts.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
import { path } from "../internal/utils/path.mjs";
/**
 * Connected account management
 */
export class ConnectedAccounts extends APIResource {
    /**
     * Initiates a new connection to an external service for a user. For OAuth-based
     * toolkits, this returns a redirect URL to complete authentication. For API
     * key-based toolkits, provide the credentials directly in the request body. Use
     * the `user_id` field to associate the connection with a specific user in your
     * system.
     *
     * **Deprecated for Composio-managed OAuth:** For Composio-managed auth configs on
     * OAuth1, OAuth2, or DCR_OAUTH schemes, this endpoint is being retired and will
     * start returning `400 BadRequest` on **2026-05-08** for new organizations and
     * **2026-07-03** for all remaining organizations. Migrate those calls to
     * `POST /api/v3/connected_accounts/link`. Custom auth configs and non-OAuth
     * schemes (API key, bearer, basic) continue to be supported here. Responses on the
     * retiring path carry a `Deprecation` header (RFC 9745) and a `Sunset` header
     * (RFC 8594) for client-side detection.
     *
     * @example
     * ```ts
     * const connectedAccount =
     *   await client.connectedAccounts.create({
     *     auth_config: { id: 'id' },
     *     connection: {},
     *   });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/v3.1/connected_accounts', { body, ...options });
    }
    /**
     * Retrieves comprehensive details of a connected account, including authentication
     * configuration, connection status, and all parameters needed for API requests.
     *
     * @example
     * ```ts
     * const connectedAccount =
     *   await client.connectedAccounts.retrieve(
     *     'ca_1a2b3c4d5e6f',
     *   );
     * ```
     */
    retrieve(nanoid, options) {
        return this._client.get(path `/api/v3.1/connected_accounts/${nanoid}`, options);
    }
    /**
     * Retrieves all connected accounts for your project. Connected accounts represent
     * authenticated user connections to external services (e.g., a user's Gmail
     * account, Slack workspace). Filter by toolkit, status, user ID, or auth config to
     * find specific connections.
     *
     * @example
     * ```ts
     * const connectedAccounts =
     *   await client.connectedAccounts.list();
     * ```
     */
    list(query = {}, options) {
        return this._client.get('/api/v3.1/connected_accounts', { query, ...options });
    }
    /**
     * Soft-deletes a connected account by marking it as deleted in the database. This
     * prevents the account from being used for API calls but preserves the record for
     * audit purposes.
     *
     * @example
     * ```ts
     * const connectedAccount =
     *   await client.connectedAccounts.delete('ca_1a2b3c4d5e6f');
     * ```
     */
    delete(nanoid, options) {
        return this._client.delete(path `/api/v3.1/connected_accounts/${nanoid}`, options);
    }
    /**
     * Update a connected account. Supports updating the alias and/or credentials. Only
     * specified fields will be updated. Set a credential field to null to remove it.
     * Alias must be unique within the same project, entity, and toolkit scope.
     *
     * @example
     * ```ts
     * const response = await client.connectedAccounts.patch(
     *   'ca_1a2b3c4d5e6f',
     *   {
     *     connection: {
     *       state: {
     *         authScheme: 'BEARER_TOKEN',
     *         val: { token: 'new_access_token' },
     *       },
     *     },
     *   },
     * );
     * ```
     */
    patch(nanoid, body = {}, options) {
        return this._client.patch(path `/api/v3.1/connected_accounts/${nanoid}`, { body, ...options });
    }
    /**
     * Initiates a new authentication flow for a connected account when credentials
     * have expired or become invalid. This may generate a new authentication URL for
     * OAuth flows or refresh tokens for other auth schemes.
     *
     * @example
     * ```ts
     * const response = await client.connectedAccounts.refresh(
     *   'ca_1a2b3c4d5e6f',
     * );
     * ```
     */
    refresh(nanoid, params = {}, options) {
        const { query_redirect_url, ...body } = params ?? {};
        return this._client.post(path `/api/v3.1/connected_accounts/${nanoid}/refresh`, {
            query: { redirect_url: query_redirect_url },
            body,
            ...options,
        });
    }
    /**
     * Updates the status of a connected account to either enabled (active) or disabled
     * (inactive). Disabled accounts cannot be used for API calls but remain in the
     * database.
     *
     * @example
     * ```ts
     * const response =
     *   await client.connectedAccounts.updateStatus(
     *     'ca_1a2b3c4d5e6f',
     *     { enabled: true },
     *   );
     * ```
     */
    updateStatus(nanoID, body, options) {
        return this._client.patch(path `/api/v3.1/connected_accounts/${nanoID}/status`, { body, ...options });
    }
}
