// @composio/core/dist/platform/node.mjs
import * as fs$1 from "node:fs";
import * as os from "node:os";
import * as path from "node:path";

//#region src/platform/node.ts
/**
* Node.js platform implementation.
* Provides full file system and OS operations using Node.js built-in modules.
*/
const platform = {
	supportsFileSystem: true,
	homedir() {
		try {
			return os.homedir();
		} catch {
			return null;
		}
	},
	joinPath(...paths) {
		return path.join(...paths);
	},
	resolvePath(...paths) {
		return path.resolve(...paths);
	},
	isAbsolutePath(filePath) {
		return path.isAbsolute(filePath);
	},
	basename(filePath) {
		return path.basename(filePath);
	},
	existsSync(filePath) {
		return fs$1.existsSync(filePath);
	},
	mkdirSync(dirPath) {
		fs$1.mkdirSync(dirPath, { recursive: true });
	},
	readFileSync(filePath, encoding) {
		if (encoding === void 0) {
			const buf = fs$1.readFileSync(filePath, { encoding: null });
			return new Uint8Array(buf.buffer, buf.byteOffset, buf.byteLength);
		}
		return fs$1.readFileSync(filePath, { encoding });
	},
	writeFileSync(filePath, content, encoding) {
		if (encoding && typeof content === "string") fs$1.writeFileSync(filePath, content, { encoding });
		else fs$1.writeFileSync(filePath, content);
	}
};

//#endregion
export { platform };
