// @composio/client/src/resources/tool-router/tool-router.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../../core/resource';
import * as SessionAPI from './session/session';
import {
  Session,
  SessionAttachParams,
  SessionAttachResponse,
  SessionConfigHistoryParams,
  SessionConfigHistoryResponse,
  SessionCreateParams,
  SessionCreateResponse,
  SessionExecuteMetaParams,
  SessionExecuteMetaResponse,
  SessionExecuteParams,
  SessionExecuteResponse,
  SessionLinkParams,
  SessionLinkResponse,
  SessionPatchParams,
  SessionPatchResponse,
  SessionProxyExecuteParams,
  SessionProxyExecuteResponse,
  SessionRetrieveResponse,
  SessionSearchParams,
  SessionSearchResponse,
  SessionToolkitsParams,
  SessionToolkitsResponse,
  SessionToolsParams,
  SessionToolsResponse,
} from './session/session';
import { APIPromise } from '../../core/api-promise';
import { RequestOptions } from '../../internal/request-options';

/**
 * (Labs) Tool router endpoints
 */
export class ToolRouter extends APIResource {
  session: SessionAPI.Session = new SessionAPI.Session(this._client);

  /**
   * Creates a new session for the tool router lab feature (Legacy). This endpoint
   * initializes a new session with specified toolkits and their authentication
   * configurations. The session provides an isolated environment for testing and
   * managing tool routing logic with scoped MCP server access.
   *
   * @example
   * ```ts
   * const response = await client.toolRouter.createSession({
   *   user_id: 'user_123456789',
   *   config: {
   *     toolkits: [
   *       {
   *         toolkit: 'gmail',
   *         auth_config_id: 'ac_1a2b3c4d5e6f',
   *       },
   *       {
   *         toolkit: 'slack',
   *         auth_config_id: 'ac_7g8h9i0j1k2l',
   *       },
   *       { toolkit: 'github' },
   *     ],
   *   },
   * });
   * ```
   */
  createSession(
    body: ToolRouterCreateSessionParams,
    options?: RequestOptions,
  ): APIPromise<ToolRouterCreateSessionResponse> {
    return this._client.post('/api/v3.1/labs/tool_router/session', { body, ...options });
  }
}

export interface ToolRouterCreateSessionResponse {
  /**
   * MCP server endpoint URL for this specific chat session
   */
  chat_session_mcp_url: string;

  /**
   * Generated session identifier (prefixed with "trs\_")
   */
  session_id: string;

  /**
   * MCP server endpoint URL for this specific tool router instance
   */
  tool_router_instance_mcp_url: string;
}

export interface ToolRouterCreateSessionParams {
  /**
   * Unique user identifier for the session owner
   */
  user_id: string;

  /**
   * Session configuration including enabled toolkits and their auth configs
   */
  config?: ToolRouterCreateSessionParams.Config;
}

export namespace ToolRouterCreateSessionParams {
  /**
   * Session configuration including enabled toolkits and their auth configs
   */
  export interface Config {
    /**
     * Whether to manually manage connections
     */
    manually_manage_connections?: boolean | null;

    /**
     * Array of toolkit configurations with optional auth configs
     */
    toolkits?: Array<Config.Toolkit>;
  }

  export namespace Config {
    export interface Toolkit {
      /**
       * Toolkit identifier (e.g., gmail, slack, github)
       */
      toolkit: string;

      /**
       * Specific auth configuration ID for this toolkit
       */
      auth_config_id?: string;
    }
  }
}

ToolRouter.Session = Session;

export declare namespace ToolRouter {
  export {
    type ToolRouterCreateSessionResponse as ToolRouterCreateSessionResponse,
    type ToolRouterCreateSessionParams as ToolRouterCreateSessionParams,
  };

  export {
    Session as Session,
    type SessionCreateResponse as SessionCreateResponse,
    type SessionRetrieveResponse as SessionRetrieveResponse,
    type SessionAttachResponse as SessionAttachResponse,
    type SessionConfigHistoryResponse as SessionConfigHistoryResponse,
    type SessionExecuteResponse as SessionExecuteResponse,
    type SessionExecuteMetaResponse as SessionExecuteMetaResponse,
    type SessionLinkResponse as SessionLinkResponse,
    type SessionPatchResponse as SessionPatchResponse,
    type SessionProxyExecuteResponse as SessionProxyExecuteResponse,
    type SessionSearchResponse as SessionSearchResponse,
    type SessionToolkitsResponse as SessionToolkitsResponse,
    type SessionToolsResponse as SessionToolsResponse,
    type SessionCreateParams as SessionCreateParams,
    type SessionAttachParams as SessionAttachParams,
    type SessionConfigHistoryParams as SessionConfigHistoryParams,
    type SessionExecuteParams as SessionExecuteParams,
    type SessionExecuteMetaParams as SessionExecuteMetaParams,
    type SessionLinkParams as SessionLinkParams,
    type SessionPatchParams as SessionPatchParams,
    type SessionProxyExecuteParams as SessionProxyExecuteParams,
    type SessionSearchParams as SessionSearchParams,
    type SessionToolkitsParams as SessionToolkitsParams,
    type SessionToolsParams as SessionToolsParams,
  };
}
