// @composio/core/dist/models/Files.node.mjs
import "../chunk-CWQeghP5.mjs";
import "../FileModifierErrors-C7z16Dfi.mjs";
import { t as telemetry } from "../Telemetry-kgWlkavX.mjs";
import { n as getFileDataAfterUploadingToS3, t as downloadFileFromS3 } from "../fileUtils.node-Dq7ebZPC.mjs";

//#region src/models/Files.node.ts
var Files = class {
	fileOptions;
	constructor(client, fileOptions = {}) {
		this.client = client;
		this.fileOptions = fileOptions;
		telemetry.instrument(this, "Files");
	}
	/**
	* Upload a file and return the file data.
	*
	* @param params - The upload parameters.
	* @param {File | string} params.file - The path to the file to upload, a URL of the file, or a File object.
	* @param {string} params.toolSlug - The slug of the tool that is uploading the file.
	* @param {string} params.toolkitSlug - The slug of the toolkit that is uploading the file.
	* @returns {Promise<FileUploadData>} The uploaded file data.
	*
	* @example
	*
	* const fileData = await composio.files.upload({
	*   file: 'path/to/file.pdf',
	*   toolSlug: 'google_drive_upload',
	*   toolkitSlug: 'google_drive'
	* });
	* */
	async upload({ file, toolSlug, toolkitSlug }) {
		return await getFileDataAfterUploadingToS3(file, {
			toolSlug,
			toolkitSlug,
			client: this.client,
			sensitiveFileUploadProtection: this.fileOptions.sensitiveFileUploadProtection,
			fileUploadPathDenySegments: this.fileOptions.fileUploadPathDenySegments
		});
	}
	/**
	* Download a file from S3 and return the file data.
	* @param s3key - The S3 key of the file to download.
	* @returns The file data.
	*/
	async download({ toolSlug, s3Url, mimeType }) {
		return await downloadFileFromS3({
			toolSlug,
			s3Url,
			mimeType,
			fileDownloadDir: this.fileOptions.fileDownloadDir
		});
	}
};

//#endregion
export { Files };
