// @composio/core/dist/utils/modifiers/FileToolModifier.node.cjs
const require_FileModifierErrors = require('../../FileModifierErrors-BYJBHiay.cjs');
const require_FileToolModifier_utils_neutral = require('../../FileToolModifier.utils.neutral-Dfwqgimk.cjs');
const require_fileUtils_node = require('../../fileUtils.node-DtnSRMWa.cjs');

//#region src/utils/modifiers/FileToolModifier.node.ts
/**
* Recursively walks a runtime value and its matching JSON-Schema node,
* uploading any string path whose schema node has `file_uploadable: true`.
* The function returns a **new** value with all substitutions applied;
* nothing is mutated in-place.
*/
const hydrateFiles = async (value, schema, ctx) => {
	if (schema?.file_uploadable) {
		if (typeof value !== "string" && !(value instanceof File)) return value;
		const runBeforeFileUpload = async (path, source) => {
			if (!ctx.beforeFileUpload) return path;
			const out = await ctx.beforeFileUpload({
				path,
				source,
				toolSlug: ctx.toolSlug,
				toolkitSlug: ctx.toolkitSlug
			});
			if (out === false) throw new require_FileModifierErrors.ComposioFileUploadAbortedError("File upload was aborted because beforeFileUpload returned false.");
			return out;
		};
		if (typeof value === "string") {
			const pathOrUrl = await runBeforeFileUpload(value, value.startsWith("http") ? "url" : "path");
			require_FileModifierErrors.logger_default.debug(`Uploading file "${pathOrUrl}"`);
			return require_fileUtils_node.getFileDataAfterUploadingToS3(pathOrUrl, {
				toolSlug: ctx.toolSlug,
				toolkitSlug: ctx.toolkitSlug,
				client: ctx.client,
				sensitiveFileUploadProtection: ctx.sensitiveFileUploadProtection,
				fileUploadPathDenySegments: ctx.fileUploadPathDenySegments,
				fileUploadAllowlist: ctx.fileUploadAllowlist
			});
		}
		if (ctx.beforeFileUpload) {
			const out = await ctx.beforeFileUpload({
				path: value.name,
				source: "file",
				toolSlug: ctx.toolSlug,
				toolkitSlug: ctx.toolkitSlug
			});
			if (out === false) throw new require_FileModifierErrors.ComposioFileUploadAbortedError("File upload was aborted because beforeFileUpload returned false.");
			if (typeof out === "string" && out !== value.name) {
				require_FileModifierErrors.logger_default.debug(`Uploading file from path "${out}" (replaced File: ${value.name})`);
				return require_fileUtils_node.getFileDataAfterUploadingToS3(out, {
					toolSlug: ctx.toolSlug,
					toolkitSlug: ctx.toolkitSlug,
					client: ctx.client,
					sensitiveFileUploadProtection: ctx.sensitiveFileUploadProtection,
					fileUploadPathDenySegments: ctx.fileUploadPathDenySegments,
					fileUploadAllowlist: ctx.fileUploadAllowlist
				});
			}
		}
		require_FileModifierErrors.logger_default.debug(`Uploading file "${value.name}"`);
		return require_fileUtils_node.getFileDataAfterUploadingToS3(value, {
			toolSlug: ctx.toolSlug,
			toolkitSlug: ctx.toolkitSlug,
			client: ctx.client,
			sensitiveFileUploadProtection: ctx.sensitiveFileUploadProtection,
			fileUploadPathDenySegments: ctx.fileUploadPathDenySegments
		});
	}
	const schemaVariants = [
		...schema?.anyOf ?? [],
		...schema?.oneOf ?? [],
		...schema?.allOf ?? []
	];
	if (schemaVariants.length > 0) {
		const firstUploadableVariant = schemaVariants.find(require_FileToolModifier_utils_neutral.schemaHasFileUploadable);
		if (firstUploadableVariant) return hydrateFiles(value, firstUploadableVariant, ctx);
	}
	if (schema?.type === "object" && schema.properties && require_FileToolModifier_utils_neutral.isPlainObject(value)) {
		const transformed = {};
		for (const [k, v] of Object.entries(value)) transformed[k] = await hydrateFiles(v, schema.properties[k], ctx);
		return transformed;
	}
	if (schema?.type === "array" && schema.items && Array.isArray(value)) {
		const itemSchema = Array.isArray(schema.items) ? schema.items[0] : schema.items;
		return Promise.all(value.map((item) => hydrateFiles(item, itemSchema, ctx)));
	}
	return value;
};
/**
* Downloads a file from S3 and returns a replacement object.
*/
const downloadS3File = async (value, ctx) => {
	const { s3url, mimetype } = value;
	try {
		require_FileModifierErrors.logger_default.debug(`Downloading from S3: ${s3url}`);
		const dl = await require_fileUtils_node.downloadFileFromS3({
			toolSlug: ctx.toolSlug,
			s3Url: s3url,
			mimeType: mimetype ?? "application/octet-stream",
			fileDownloadDir: ctx.fileDownloadDir
		});
		require_FileModifierErrors.logger_default.debug(`Downloaded → ${dl.filePath}`);
		return {
			uri: dl.filePath,
			file_downloaded: dl.filePath ? true : false,
			s3url,
			mimeType: dl.mimeType
		};
	} catch (err) {
		require_FileModifierErrors.logger_default.error(`Download failed: ${s3url}`, { cause: err });
		return {
			uri: "",
			file_downloaded: false,
			s3url,
			mimeType: mimetype ?? "application/octet-stream"
		};
	}
};
/**
* Recursively walks an arbitrary value and its matching JSON-Schema node.
* Whenever it encounters an object that represents a file reference
* (i.e. has an `s3url`), it downloads the file and returns a replacement:
*   {
*     uri: "<local-path>",
*     file_downloaded: true | false,
*     s3url: "<original S3 URL>",
*     mimeType: "<detected-or-fallback-mime-type>"
*   }
*
* The function is side-effect-free: it never mutates the input value.
*/
const hydrateDownloads = async (value, schema, ctx) => {
	if (require_FileToolModifier_utils_neutral.isPlainObject(value) && typeof value.s3url === "string") return downloadS3File(value, ctx);
	if (schema?.file_downloadable && require_FileToolModifier_utils_neutral.isPlainObject(value) && typeof value.s3url === "string") return downloadS3File(value, ctx);
	const schemaVariants = [
		...schema?.anyOf ?? [],
		...schema?.oneOf ?? [],
		...schema?.allOf ?? []
	];
	if (schemaVariants.length > 0) {
		const downloadableVariants = schemaVariants.filter(require_FileToolModifier_utils_neutral.schemaHasFileDownloadable);
		let result = value;
		for (const variant of downloadableVariants) result = await hydrateDownloads(result, variant, ctx);
		if (downloadableVariants.length === 0) return hydrateDownloads(value, void 0, ctx);
		return result;
	}
	if (require_FileToolModifier_utils_neutral.isPlainObject(value)) {
		const pairs = await Promise.all(Object.entries(value).map(async ([k, v]) => [k, await hydrateDownloads(v, schema?.properties?.[k], ctx)]));
		return Object.fromEntries(pairs);
	}
	if (Array.isArray(value)) {
		const itemSchema = schema?.items ? Array.isArray(schema.items) ? schema.items[0] : schema.items : void 0;
		return Promise.all(value.map((item) => hydrateDownloads(item, itemSchema, ctx)));
	}
	return value;
};
var FileToolModifier = class {
	client;
	fileUploadPathOptions;
	constructor(client, fileUploadPathOptions = {}) {
		this.client = client;
		this.fileUploadPathOptions = fileUploadPathOptions;
	}
	async modifyToolSchema(schema) {
		if (!schema.inputParameters?.properties) return schema;
		const properties = require_FileToolModifier_utils_neutral.transformProperties(schema.inputParameters.properties);
		return {
			...schema,
			inputParameters: {
				...schema.inputParameters,
				properties
			}
		};
	}
	async fileUploadModifier(tool, options) {
		const { params, toolSlug, toolkitSlug = "unknown" } = options;
		const { arguments: args } = params;
		if (!args || typeof args !== "object") return params;
		try {
			const newArgs = await hydrateFiles(args, tool.inputParameters, {
				toolSlug,
				toolkitSlug,
				client: this.client,
				...this.fileUploadPathOptions
			});
			return {
				...params,
				arguments: newArgs
			};
		} catch (error) {
			if (error instanceof require_FileModifierErrors.ComposioSensitiveFilePathBlockedError || error instanceof require_FileModifierErrors.ComposioFileUploadAbortedError || error instanceof require_FileModifierErrors.ComposioFileUploadPathNotAllowedError || error instanceof require_FileModifierErrors.ComposioFileNotFoundError) throw error;
			throw new require_FileModifierErrors.ComposioFileUploadError("Failed to upload file", { cause: error });
		}
	}
	async fileDownloadModifier(tool, options) {
		const { result, toolSlug } = options;
		const dataWithDownloads = await hydrateDownloads(result.data, tool.outputParameters, {
			toolSlug,
			fileDownloadDir: this.fileUploadPathOptions.fileDownloadDir
		});
		return {
			...result,
			data: dataWithDownloads
		};
	}
};

//#endregion
exports.FileToolModifier = FileToolModifier;
